core/store/src/lib.rs

```diff
@@ -0,0 +1,180 @@
+        let transaction = storage.transaction();
+        StoreUpdate { storage, transaction, trie: Some(trie) }
+    }
+
+    pub fn set(&mut self, column: Option<u32>, key: &[u8], value: &[u8]) {
+        self.transaction.put(column, key, value)
+    }
+
+    pub fn set_ser<T: Encode>(
+        &mut self,
+        column: Option<u32>,
+        key: &[u8],
+        value: &T,
+    ) -> Result<(), io::Error> {
+        let data = Encode::encode(value)?;
+        self.set(column, key, &data);
+        Ok(())
+    }
+
+    pub fn delete(&mut self, column: Option<u32>, key: &[u8]) {
+        self.transaction.delete(column, key);
+    }
+
+    /// Merge another store update into this one.
+    pub fn merge(&mut self, other: StoreUpdate) {
+        if self.trie.is_none() {
+            if let Some(trie) = other.trie {
+                self.trie = Some(trie);
+            }
+        }
+        self.merge_transaction(other.transaction);
+    }
+
+    /// Merge DB Transaction.
+    pub fn merge_transaction(&mut self, transaction: DBTransaction) {
+        for op in transaction.ops {
+            match op {
+                DBOp::Insert { col, key, value } => self.transaction.put(col, &key, &value),
+                DBOp::Delete { col, key } => self.transaction.delete(col, &key),
+            }
+        }
+    }
+
+    pub fn commit(self) -> Result<(), io::Error> {
+        if let Some(trie) = self.trie {
+            trie.update_cache(&self.transaction)?;
+        }
+        self.storage.write(self.transaction)
+    }
+}
+
+impl fmt::Debug for StoreUpdate {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Store Update {{\n")?;
+        for op in self.transaction.ops.iter() {
+            match op {
+                DBOp::Insert { col, key, value: _ } => {
+                    write!(f, "  + {:?} {}\n", col, to_base64(key))?
+                }
+                DBOp::Delete { col, key } => write!(f, "  - {:?} {}\n", col, to_base64(key))?,
+            }
+        }
+        write!(f, "}}\n")
+    }
+}
+
+pub fn read_with_cache<'a, T: Decode + DeserializeOwned + 'a>(
+    storage: &Store,
+    col: Option<u32>,
+    cache: &'a mut SizedCache<Vec<u8>, T>,
+    key: &[u8],
+) -> io::Result<Option<&'a T>> {
+    let key_vec = key.to_vec();
+    if cache.cache_get(&key_vec).is_some() {
+        return Ok(Some(cache.cache_get(&key_vec).unwrap()));
+    }
+    if let Some(result) = storage.get_ser(col, key)? {
+        cache.cache_set(key.to_vec(), result);
+        return Ok(cache.cache_get(&key_vec));
+    }
+    Ok(None)
+}
+
+pub fn create_store(path: &str) -> Arc<Store> {
+    let db_config = DatabaseConfig::with_columns(Some(NUM_COLS));
+    let db = Arc::new(Database::open(&db_config, path).expect("Failed to open the database"));
+    Arc::new(Store::new(db))
+}
+
+/// Reads an object from Trie.
+pub fn get<T: DeserializeOwned>(state_update: &TrieUpdate, key: &[u8]) -> Option<T> {
```

Comment:
Ideally getting stuff from state update should not require deserialization.

Reply:
? just call `state_update.get(key)` then